### PR TITLE
FileTransfer: add variant for Directory in the Offer type

### PR DESCRIPTION
**Pull request by vu3rdd**
Adds type variant for Directory offer, json encode/decode instances and a round trip test.

### CODE REVIEW

File: src/MagicWormhole/Internal/FileTransfer.hs [hunk 1]
```diff
@@ -29,12 +29,26 @@ import System.Posix.Types (FileOffset)
...
+    Text
+    -- ^ Mode. Currently always "zipfile/deflated".
+    Text
+    -- ^ Directory Name.
+    Integer
+    -- ^ size of the transmitted compressed data in bytes
+    Integer
+    -- ^ estimated total size of the uncompressed directory
+    Integer
+    -- ^ number of files and directories being sent
```
Comment (jml) [inline src/MagicWormhole/Internal/FileTransfer.hs]: Personally, I think this is getting big enough that the fields should be named.

File: src/MagicWormhole/Internal/FileTransfer.hs [hunk 2]
```diff
@@ -29,12 +29,26 @@ import System.Posix.Types (FileOffset)
 data Offer
   -- | A simple text message.
   = Message Text
+  -- | Offer a File with filename and size.
   | File FilePath FileOffset
+  -- | Offer a Directory
+  | Directory
+    Text
+    -- ^ Mode. Currently always "zipfile/deflated".
```
Comment (jml) [inline src/MagicWormhole/Internal/FileTransfer.hs]: I don't know enough about this layer of Magic Wormhole, but is this really an arbitrary unicode string? Would this be better encoded as a mime type or enum or something?

File: src/MagicWormhole/Internal/FileTransfer.hs [hunk 3]
```diff
@@ -29,12 +29,26 @@ import System.Posix.Types (FileOffset)
...
+    Text
+    -- ^ Directory Name.
+    Integer
+    -- ^ size of the transmitted compressed data in bytes
+    Integer
+    -- ^ estimated total size of the uncompressed directory
+    Integer
```
Comment (jml) [inline src/MagicWormhole/Internal/FileTransfer.hs]: Again, lack of detailed knowledge of the protocol is preventing me from being authoritative: are these all really unbounded? Would, e.g., `Int64` be more appropriate?

**Comment by vu3rdd** [comment]
@jml The reason it did not catch the incomplete pattern is because have only been using the library. In deed, it is an error to have incomplete patterns in hocus-pocus: https://github.com/LeastAuthority/haskell-magic-wormhole/blob/master/magic-wormhole.cabal#L67

I will fix your concerns.

File: src/MagicWormhole/Internal/FileTransfer.hs (hunk 2 above)
Comment (vu3rdd) [reply src/MagicWormhole/Internal/FileTransfer.hs]: Yes, it is a hardcoded arbitrary unicode string. The python client checks if the sender sends this exact string.

File: src/MagicWormhole/Internal/FileTransfer.hs (hunk 3 above)
Comment (vu3rdd) [reply src/MagicWormhole/Internal/FileTransfer.hs]: I am not aware of any upper bounds either. After establishing the tcp pipe, we send records. Each of the record itself has a size upper bound of 4GB. But I am not aware of other bounds for the zipfiles representing directories or number of files in the directories etc.

File: src/MagicWormhole/Internal/FileTransfer.hs (hunk 1 above)
Comment (jml) [reply src/MagicWormhole/Internal/FileTransfer.hs]: FWIW, I meant named fields (e.g. `numFiles :: Int64`), not named types. I'm not convinced the new type aliases add much value.

File: tests/Generator.hs [hunk 4]
```diff
@@ -106,4 +106,10 @@ offerMessages :: MonadGen m => m F.Offer
 offerMessages = Gen.choice
   [ F.Message <$> (Gen.text (Range.linear 0 1024) Gen.unicode)
   , F.File <$> toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum)  <*> (fromIntegral <$> (Gen.int (Range.linear 0 maxBound)))
+  , F.Directory
+    <$> (toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum))
+    <*> (toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum))
+    <*> (fromIntegral <$> (Gen.int (Range.linear 0 maxBound)))
+    <*> (fromIntegral <$> (Gen.int (Range.linear 0 maxBound)))
+    <*> (fromIntegral <$> (Gen.int (Range.linear 0 maxBound)))
```
Comment (exarkun) [inline tests/Generator.hs]: These three fields are `Int64` and can contain negative integers.  I think `(Gen.int (Range.linear 0 maxBound))` will only generate `[0, maxBound]`.  Maybe the types of these fields should instead be `Data.Word.Word64`?  A negative value for any of them seems nonsensical.  On the other hand, I don't like the overflow behavior of `Word64`.  On the first hand again, `Int64` has the same behavior!  So `Word64` might still be slightly better.  But I wonder what the Aeson behavior of decoding "-1"` into a Word64 is.  An error?  A overflow?

So unfortunately I'm not really sure what I'm suggesting here.  Apart from an idea that the generator for the type should exercise the full domain of the type (particularly the crazy parts).

Comment (exarkun) [reply tests/Generator.hs]: Natural numbers are apparently a thing.

http://hackage.haskell.org/package/base-4.11.1.0/docs/Numeric-Natural.html

and also

http://hackage.haskell.org/package/naturals-0.2.0.2/docs/Data-Natural.html

Comment (vu3rdd) [reply tests/Generator.hs]: > But I wonder what the Aeson behavior of decoding "-1"` into a Word64 is. An error? A overflow?

I am not sure either. It definitely seem wrong to use a negative number for these fields as they all represent sizes. `Numeric.Natural` is nice.

File: tests/Generator.hs [hunk 5]
```diff
@@ -106,4 +106,10 @@ offerMessages :: MonadGen m => m F.Offer
 offerMessages = Gen.choice
   [ F.Message <$> (Gen.text (Range.linear 0 1024) Gen.unicode)
   , F.File <$> toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum)  <*> (fromIntegral <$> (Gen.int (Range.linear 0 maxBound)))
+  , F.Directory
+    <$> (toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum))
+    <*> (toS <$> (Gen.text (Range.linear 0 100) Gen.alphaNum))
```
[the remark this answers is not in the document]
Comment (exarkun) [reply tests/Generator.hs]: > I will change that. dirName, like the filename only contains the name and not the path, so it shouldn't have the . and /.

Does anything _prevent_ it from having this, though?  What if I write a malicious client that sends a `dirName` with path separators or an absolute path?

[the remark this answers is not in the document]
Comment (vu3rdd) [reply tests/Generator.hs]: Yes, that is definitely possible. So, I guess we should not make assumptions here on what the string would contain and handle it in the elsewhere (client?) where we interpret the name and take only the file/directory name and discard the base path perhaps?

**Comment by vu3rdd** [comment]
@exarkun Except for the `DirectoryMode` suggestion, which I will work on separately, I have tried to address your comments.

**Comment by vu3rdd** [comment]
@exarkun just pushed a commit to the PR to encode the directory mode into a type with proper serialization.

**Comment by exarkun** [comment]
These latest changes look good to me.

**Comment by vu3rdd** [comment]
> These latest changes look good to me.

Thanks. Merging it.

**Pull request merged**